devito/passes/iet/misc.py

```diff
@@ -104,9 +113,89 @@ def relax_incr_dimensions(iet, **kwargs):
+            # IncrDimensions as a rule to generate assumptions such as:
+            # e.g. for 2-level blocking [x (inner Incr), x0_blk1, x0_blk0, x (outer root)]
+            # [x0_blk1 > x0_blk0, x0_blk1_size > x0_blk0_size]
+            # `defines` keeps all the ancestors and descendants of `i.dim`.
+            defines = sorted(i.dim._defines, key=lambda x: (not x.is_Incr or -x._depth))
+
+            # Collect assumptions based on hierarchy
+            acs = [j for j in defines if j is not i.dim and j.is_Incr]
+            assumptions = as_list(chain.from_iterable((Gt(j.step, k.step), Gt(j, k))
+                                  for j, k in product(acs, acs) if j._depth > k._depth))
+
+            # Step 3: Reduce defines min/max candidates using assumptions on `defines`
+            defmin = reduce_relation(max, defines, assumptions)
+            defmax = reduce_relation(min, defines, assumptions)
+
+            # Collect symbolic_min, symbolic_max from defmin/defmax lists
+            defmin = [j.symbolic_min for j in defmin]
+            defmax = [j.symbolic_max for j in defmax]
+
+            # Drop duplicates from defmin/defmax and keep them Ordered
+            defmin = list(OrderedDict.fromkeys(defmin))
+            defmax = list(OrderedDict.fromkeys(defmax))
+
+            # Step 4: Drop a candidate if it is the symbolic_min/max of symbol candidates
+            for k in defmin.copy():  # TOFIX: Should not use a copy I guess?
+                try:
+                    if k.symbolic_min in defmin:
+                        defmin.remove(k.symbolic_min)
+                except:
+                    pass
+
+            for k in defmax.copy():
+                try:
+                    if k.symbolic_max in defmax:
+                        defmax.remove(k.symbolic_max)
+                except:
+                    pass
+
+            # At this point the candidates for the start and end of the `Iteration`
+            # comprise of candidates stemming (i) from `i.dim` and (ii) candidates
+            # stemming from other dims, (in space blocking other dims is usually root.)
+            # The canidates stemming from 'i.dim' can be compared and for those we need
+            # the minimum of them for the start and the maximum of them for the end so
+            # as to span the whole extent of the Iteration's Dimension (i.dim iteration
+            # extent).
+            # We split on comparable and uncomparable. Comparable ones can be further
+            # simplfied, (subdimensions, offsets incurred) and non-comparable ones
+            # include candidates that cannot be simplified more (no clear rule of
+            # further simplification exists)
+            # For the start: MAX(uncomparable, MIN(comparable)) --> MAX(uncomparable)
+            # For the end: MIN(uncomparable, MAX(comparable)) --> MIN(uncomparable)
+            defmin = reduce_relation(min, defmin, assumptions=None)
```

Comment:
I still don't understand why we need 2x2 calls to `reduce_relation` rather than 2x1 (ie once for MIN and once for MAX)

Please write example